File: internal/bridge/nsq.go
package bridge

import (
	"context"

	nsqservice "github.com/amigoer/mq-studio/internal/service/nsq"
)

// NSQService is the renderer's entry point for the operations only NSQ has.
// Topics, channels and publishing all go through the canonical services; what
// is here is the rest.
type NSQService struct {
	service *nsqservice.Service
}

// NewNSQService wires the bridge to the service.
func NewNSQService(service *nsqservice.Service) *NSQService {
	return &NSQService{service: service}
}

// CreateTopic declares a topic on every nsqd in the connection.
//
// Not TopicService.Create, whose input is a broker address, two queue counts
// and a permission string - RocketMQ's vocabulary, of which an NSQ topic has
// none. A name is the whole of what one is.
func (s *NSQService) CreateTopic(connID int, name string) error {
	return s.service.CreateTopic(context.Background(), connID, name)
}

// RemoveTopic deletes a topic, its channels and its registration in the
// discovery tier. There is no undo.
func (s *NSQService) RemoveTopic(connID int, name string) error {
	return s.service.RemoveTopic(context.Background(), connID, name)
}

// EmptyTopic discards what the topic and every channel under it are holding.
func (s *NSQService) EmptyTopic(connID int, name string) error {
	return s.service.EmptyTopic(context.Background(), connID, name)
}

// SetTopicPaused stops or resumes delivery into a topic's channels.
//
// A boolean rather than two methods, because it is one control with two
// positions and the page reads the current state back off the topic listing.
func (s *NSQService) SetTopicPaused(connID int, name string, paused bool) error {
	return s.service.SetTopicPaused(context.Background(), connID, name, paused)
}

// NSQChannelInput names a channel, which takes two fields because a channel
// belongs to a topic: "analytics" under two topics is two channels with
// nothing in common.
type NSQChannelInput struct {
	Topic   string `json:"topic"`
	Channel string `json:"channel"`
}

// CreateChannel declares a channel on every nsqd carrying its topic.
//
// There is no position to start it from. What it gets is whatever nsqd is
// still holding: nothing, on a topic that already has a channel, and the
// topic's own queue on one that had none to copy into.
func (s *NSQService) CreateChannel(connID int, input NSQChannelInput) error {
	return s.service.CreateChannel(context.Background(), connID, input.Topic, input.Channel)
}

// RemoveChannel deletes a channel and its backlog, in the discovery tier as
// well as on every nsqd. There is no undo.
func (s *NSQService) RemoveChannel(connID int, input NSQChannelInput) error {
	return s.service.RemoveChannel(context.Background(), connID, input.Topic, input.Channel)
}

// EmptyChannel discards one channel's backlog and leaves the rest of the topic
// alone, which is what separates it from emptying the topic.
func (s *NSQService) EmptyChannel(connID int, input NSQChannelInput) error {
	return s.service.EmptyChannel(context.Background(), connID, input.Topic, input.Channel)
}

// SetChannelPaused stops or resumes delivery to one channel's consumers. The
// other channels under the topic keep running.
func (s *NSQService) SetChannelPaused(connID int, input NSQChannelInput, paused bool) error {
	return s.service.SetChannelPaused(
		context.Background(), connID, input.Topic, input.Channel, paused)
}

File: frontend/src/api/nsq.ts
import { NSQService } from "@bindings/bridge";

/**
 * The NSQ-only half of the surface.
 *
 * Reading topics is not here: a topic is a destination, and api/topic.ts
 * already answers the whole read side. What this carries is what the canonical
 * services cannot express - a create with no vocabulary but a name, and a
 * pause, which no other family in this app has.
 */

/**
 * Declare a topic on every nsqd in the connection.
 *
 * Not through the canonical create: that one collects a broker address, two
 * queue counts and a permission string, and an NSQ topic has none of them. It
 * is created everywhere rather than on one daemon because a producer that
 * connected to another would otherwise auto-create it later with none of its
 * channels.
 */
export const createTopic = (connID: number, name: string): Promise<void> =>
  NSQService.CreateTopic(connID, name);

/**
 * Delete a topic, its channels and its registration in the discovery tier.
 *
 * The last part is not decoration: nsqd forgets a deleted topic and
 * nsqlookupd does not, so a delete that stopped at nsqd leaves the name where
 * a consumer looking it up still finds it.
 */
export const removeTopic = (connID: number, name: string): Promise<void> =>
  NSQService.RemoveTopic(connID, name);

/**
 * Discard everything the topic and every channel under it are holding.
 *
 * The channels are the point. nsqd copies each message into every channel as
 * it arrives, and emptying the topic alone touches only its own queue - which
 * on a topic anything is consuming is already zero.
 */
export const emptyTopic = (connID: number, name: string): Promise<void> =>
  NSQService.EmptyTopic(connID, name);

/**
 * Stop or resume delivery into a topic's channels.
 *
 * Publishing carries on while a topic is paused: what stops is the copy into
 * each channel, so the messages pile up in the topic itself rather than being
 * refused.
 */
export const setTopicPaused = (connID: number, name: string, paused: boolean): Promise<void> =>
  NSQService.SetTopicPaused(connID, name, paused);
